src/replication.c

```diff
@@ -126,37 +126,215 @@ static int handle_mgmt_event_fd(replica_t *replica);
+		mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+		spec->rebuild_replica_count++;  
+	
+		MTX_LOCK(&healthy_replica->r_mtx);
+		TAILQ_INSERT_TAIL(&healthy_replica->mgmt_cmd_queue, mgmt_cmd,
+		    mgmt_cmd_next);
+		MTX_UNLOCK(&healthy_replica->r_mtx);
+
+		if (write(healthy_replica->mgmt_eventfd1, &num,
+		    sizeof (num)) != sizeof (num)) {
+			REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+			    "for replica(%p)\n", replica);
+			ret = -1;
+		}
+		return ret;
+	}
+
+	// Case 2
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if (replica == target_replica)
+			continue;
+
+		spec->rebuild_replica_count++;  
+	
+		mgmt_cmd = malloc(sizeof (mgmt_cmd_t));
+		BUILD_REPLICA_MGMT_HDR(rmgmtio, mgmt_opcode, data_len);
+
+		mgmt_cmd->io_hdr = rmgmtio;
+		mgmt_cmd->io_bytes = 0;
+		mgmt_cmd->data = (char *)malloc(data_len);
+		snprintf((char *)mgmt_cmd->data, data_len, "%s", spec->volname);
+		mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+		MTX_LOCK(&replica->r_mtx);
+		TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd,
+		    mgmt_cmd_next);
+		MTX_UNLOCK(&replica->r_mtx);
+		if (write(replica->mgmt_eventfd1, &num, sizeof (num)) !=
+		    sizeof (num)) {
+			REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+			    "for replica(%p)\n", replica);
+			return (ret = -1);
+		}
+	}
+	return ret;
+}
+
+static int
+start_rebuild(void *buf, replica_t *replica, uint64_t data_len)
+{
+	
+	int ret = 0;
+	uint64_t num = 1;
+	zvol_io_hdr_t *rmgmtio = NULL;
+	zvol_op_code_t mgmt_opcode = ZVOL_OPCODE_START_REBUILD;
+	mgmt_cmd_t *mgmt_cmd;
+
+	mgmt_cmd = malloc(sizeof (mgmt_cmd_t));
+	BUILD_REPLICA_MGMT_HDR(rmgmtio, mgmt_opcode, data_len);
+
+	mgmt_cmd->io_hdr = rmgmtio;
+	mgmt_cmd->io_bytes = 0;
+	mgmt_cmd->data = buf;
+	mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+	MTX_LOCK(&replica->r_mtx);
+	TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd, mgmt_cmd_next);
+	MTX_UNLOCK(&replica->r_mtx);
+
+	if (write(replica->mgmt_eventfd1, &num, sizeof (num)) !=
+	    sizeof (num)) {
+		REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+		    "for replica(%p)\n", replica);
+		return (ret = -1);
+	}
+	REPLICA_LOG("Start_rebuild opcode sent for Replica ip:%s port:%d "
+	    "state:%d\n", replica->ip, replica->port, replica->state);
+	return ret;
+}
+
+/* Rebuild can be of two type:-
+ * 1. Rebuilding a replica from healthy replica
+ * 2. Rebuilding a replica from all other replica(Mesh rebuild)
+ * API does the following tasks:-
+ * 1. find if there is healthy replica available for rebuilding
+ * 2. find downgraded replica for rebuild
+ * 3. Call prepare_for_rebuild to on healthy replica or on all
+ * other replicas to get IP address and port for replicas.
  */
 void
 update_volstate(spec_t *spec)
 {
+	int ret = 0;
 	uint64_t max;
-	replica_t *replica;
+	replica_t *replica = NULL;
+	replica_t *target_replica = NULL;
+	replica_t *healthy_replica = NULL;
+
+	if (spec->rebuild_in_progress == true) {
```

Comment:
returning here will lead to issue in not updating spec's ready state. It would be better to move rebuilding functionality to a different function, and let update_volstate to update spec's ready state.

Reply:
ok.. lets change this to assert

Reply:
Done.